# ***Predicción de la accidentalidad en Medellin***

Contexto del problema:

Objetivo:

Métodos empleados y objetivos de desarrollo:


# ***1. Previsualización de datos***

AÑO = Año de ocurrencia del incidente	true

CBML	= Codigo catastral que corresponde al codigo comuna, barrio, manzana, lote catastral de un predio.

CLASE_ACCIDENTE = Clasificación del IPAT sobre la clase de accidente de transito: choque, atropello, volcamiento, caida de ocupante, incendio, u otro (que no corresponde a las anteriores 5 clasificaciones, p. ej: sumersión)

DIRECCION	= Dirección donde ocurrió el incidente.

DIRECCION ENCASILLADA = Dirección encasillada que entrega el geocodificador.

DISEÑO = Sitio de la vía donde ocurrió el accidente: Cicloruta, Glorieta, Interseccion, Lote o Predio, Paso a Nivel, Paso Elevado, Paso Inferior, Pontón, Puente, Tramo de via, Tunel, Via peatonal.

EXPEDIENTE = Consecutivo que asigna UNE, según el orden de llegada de los expedientes para su diligenciamiento.

FECHA_ACCIDENTE = Fecha del accidente, proviene del IPAT - Informe Policial de accidente de Tránsito.

FECHA_ACCIDENTES = Fecha de los accidente (formato YYYY-MM-DD hh:mi:ss), proviene del IPAT - Informe Policial de accidentes de Tránsito.

GRAVEDAD_ACCIDENTE = Clasificación del IPAT - Informe Policial de Accidentes de Tránsito, sobre la gravedad del accidente, corresponde al resultado más grave presentado en el accidente. Daños materiales "Sólo daños", accidente con heridos "Herido", accidente con muertos "Muerto". No indica cantidad.

MES = Mes de ocurrencia del incidente vial.

NRO_RADICADO = Consecutivo que asigna UNE, según el orden de llegada de los expedientes para su diligenciamient.

NUMCOMUNA = Numero de la comununa en la que ocurrio incidente vial.

BARRIO = Barrio de ocurrencia del incidente vial.

COMUNA = Denominación con la cual se identifica cada Comuna o Corregimiento. 01:Popular 02:Santa Cruz 03:Manrique 04:Aranjuez 05:Castilla 06:Doce de Octubre 07:Robledo 08:Villa Hermosa 09:Buenos Aires 10:La Candelaria 11:Laureles - Estadio 12:La América 13:San Javier 14:El Poblado 15:Guayabal 16:Belén 50:San Sebastián de Palmitas 60:San Cristobal 70:Altavista 80:San Antonio de Prado 90:Santa Elena 99:Toda la Ciudad.

LOCATION = Fuente de información con la cual se realizó la geocodificación. [longitud, latitud].

X = Coordenada X en metros del accidente, en sistema de coordenadas MAGNA Medellin Local.

Y = Coordenada Y en metros del accidente, en sistema de coordenadas MAGNA Medellin Local.

In [104]:
#Librerias a utilizar
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# Statistics & Mathematics
import scipy.stats as stats
import statsmodels.api as sm

# Preprocessing data
from sklearn.preprocessing import MinMaxScaler, RobustScaler, StandardScaler, QuantileTransformer, FunctionTransformer

# Model Selection for Cross Validation
from sklearn.model_selection import StratifiedKFold, KFold, train_test_split,TimeSeriesSplit

# Machine Learning metrics
from sklearn import metrics
from sklearn.neighbors import KNeighborsClassifier

# ML classifiers
from sklearn.ensemble import (
    HistGradientBoostingClassifier, AdaBoostClassifier,
    RandomForestClassifier, GradientBoostingClassifier,
    StackingClassifier, VotingClassifier
    )
from sklearn.linear_model import LogisticRegression
# Randomizer
import random
# Encoder of categorical variables
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder

In [105]:
pd.options.mode.chained_assignment = None
import warnings
warnings.filterwarnings("ignore", category=UserWarning)

# ***2. Visualización de los datos***

In [106]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [107]:
#Cargamos la base de datos, le indicamos la ruta en donde se encuentra
data = pd.read_csv('/content/drive/Shareddrives/FdeAnaliticaG4/Proyecto2/incidentes_viales.csv', sep=';', low_memory=False)
data

AÑO  CBML CLASE_ACCIDENTE      DIRECCION       DIRECCION ENCASILLADA  \
0       2015  1013  Caida Ocupante    CR 46 CL 43    CR  046   043  000 00000   
1       2015  1020          Choque  CR 43 A CL 29  CR  043 A   029  000 00000   
2       2015  0205            Otro    CR 46 CL 37    CR  046   037  000 00000   
3       2015  0101       Atropello   CR 34 CL 112    CR  034   112  000 00000   
4       2015  1504          Choque    CR 50 CL 14    CR  050   014  000 00000   
...      ...   ...             ...            ...                         ...   
270760  2017  0913       Atropello  CR 28 CL 38 F  CR  028   038 F  000 00000   
270761  2017  1115  Caida Ocupante    CL 53 CR 69    CL  053   069  000 00000   
270762  2017  0513          Choque  CR 64 C CL 92  CR  064 C   092  000 00000   
270763  2017  0607  Caida Ocupante  CR 78 A CL 92  CR  078 A   092  000 00000   
270764  2017  0507          Choque   CR 63 CL 110    CR  063   110  000 00000   

               DISEÑO  EXPEDIENTE      FECHA_ACCIDENTE  \
0        Tramo de via  A000259731  21/10/2015 05:58:00   
1        Tramo de via  A000261725  05/11/2015 03:00:00   
2        Tramo de via  A000259739  21/10/2015 07:40:00   
3        Tramo de via  A000259752  21/10/2015 12:00:00   
4        Tramo de via  A000259949  03/11/2015 06:16:00   
...               ...         ...                  ...   
270760  Lote o Predio  A000585344  21/04/2017 17:30:00   
270761  Lote o Predio  A000580176  10/03/2017 16:30:00   
270762   Tramo de via  A000580528  10/03/2017 17:10:00   
270763   Tramo de via  A000580245  10/03/2017 20:00:00   
270764   Tramo de via  A000580424  10/03/2017 22:40:00   

                FECHA_ACCIDENTES GRAVEDAD_ACCIDENTE  MES NRO_RADICADO  \
0       2015-10-21T10:58:00.000Z        Con heridos   10      1508668   
1       2015-11-05T08:00:00.000Z         Solo daños   11      1510621   
2       2015-10-21T12:40:00.000Z        Con heridos   10      1508691   
3       2015-10-21T17:00:00.000Z        Con heridos   10      1508694   
4       2015-11-03T11:16:00.000Z         Solo daños   11      1510634   
...                          ...                ...  ...          ...   
270760  2017-04-21T22:30:00.000Z        Con heridos    4      1579271   
270761  2017-03-10T21:30:00.000Z        Con heridos    3      1573925   
270762  2017-03-10T22:10:00.000Z         Solo daños    3      1574088   
270763  2017-03-11T01:00:00.000Z        Con heridos    3      1573855   
270764  2017-03-11T03:40:00.000Z         Solo daños    3      1574173   

       NUMCOMUNA                     BARRIO            COMUNA  \
0             10               Barrio Colón     La Candelaria   
1             10                  San Diego     La Candelaria   
2             02                 La Francia        Santa Cruz   
3             01  Santo Domingo Savio No. 1           Popular   
4             15                   Santa Fé          Guayabal   
...          ...                        ...               ...   
270760        09                     Loreto      Buenos Aires   
270761        11             Cuarta Brigada  Laureles Estadio   
270762        05      Francisco Antonio Zea          Castilla   
270763        06                    Kennedy   Doce de Octubre   
270764        05          Héctor Abad Gómez          Castilla   

                               LOCATION          X           Y  
0       [-75.5688011014, 6.24312304123]  834949.69  1182357.34  
1       [-75.5693883283, 6.22870030622]  834880.17  1180762.02  
2       [-75.5503911403, 6.29869502502]  837004.94  1188499.17  
3       [-75.5442195069, 6.30019383306]  837688.53  1188663.05  
4        [-75.578180998, 6.21699243701]  833903.11  1179469.60  
...                                 ...        ...         ...  
270760  [-75.5559460731, 6.23260034372]  836369.48  1181189.27  
270761  [-75.5851522144, 6.26044928627]  833145.10  1184279.21  
270762  [-75.5713587099, 6.28457841509]  834679.61  1186944.12  
270763  [-75.5831132045, 6.2895

# ***3. Limpieza de los datos***

Como se puede apreciar las variables que conforman el conjunto de datos en su gran mayoría son variables categóricas, para el objetivo de la aplicación algunas no son importantes:

CBML: Ya que es más útil tener el nombre del barrio o comuna que el mismo código.

DIRECCIÓN Y DIRECCIÓN ENCASILLADA: No nos van aportar mucho, ya que para eso tenemos las coordenadas geograficas.

EXPEDIENTE: Es solo un identificador, que tampoco va aportar al uso de los datos.

FECHA_ACCIDENTE: De esta variable solo se va extraer el día, ya que tanto el mes como el año se tienen en variables aparte, teniendo en cuenta que la minima unidad temporal va ser el día, por tanto se puede descartar la hora, min, etc.

FECHA_ACCIDENTES: Ya se tienen los valores importantes a usar.

NRO_RADICADO: Es otro identificador más.

X y Y: Se van a utilizar las coordenadas de longitud y latitud.

COMUNA: En los datos vemos que existen dos variables una para la comuna y otro para el número de la columna, por lo que se va suprimir la variable del nombre de la columna, entendiendo que ya se ve representada por el número.

La variable LOCATION se va transformar en dos, una para longitud que va tener el primer valor, y otra para la variable latitud que va tener el segundo valor de la lista.

In [108]:
data = data.drop(['CBML','DIRECCION ENCASILLADA','EXPEDIENTE','FECHA_ACCIDENTES','NRO_RADICADO','X','Y','COMUNA'], axis=1)
data

AÑO CLASE_ACCIDENTE      DIRECCION         DISEÑO  \
0       2015  Caida Ocupante    CR 46 CL 43   Tramo de via   
1       2015          Choque  CR 43 A CL 29   Tramo de via   
2       2015            Otro    CR 46 CL 37   Tramo de via   
3       2015       Atropello   CR 34 CL 112   Tramo de via   
4       2015          Choque    CR 50 CL 14   Tramo de via   
...      ...             ...            ...            ...   
270760  2017       Atropello  CR 28 CL 38 F  Lote o Predio   
270761  2017  Caida Ocupante    CL 53 CR 69  Lote o Predio   
270762  2017          Choque  CR 64 C CL 92   Tramo de via   
270763  2017  Caida Ocupante  CR 78 A CL 92   Tramo de via   
270764  2017          Choque   CR 63 CL 110   Tramo de via   

            FECHA_ACCIDENTE GRAVEDAD_ACCIDENTE  MES NUMCOMUNA  \
0       21/10/2015 05:58:00        Con heridos   10        10   
1       05/11/2015 03:00:00         Solo daños   11        10   
2       21/10/2015 07:40:00        Con heridos   10        02   
3       21/10/2015 12:00:00        Con heridos   10        01   
4       03/11/2015 06:16:00         Solo daños   11        15   
...                     ...                ...  ...       ...   
270760  21/04/2017 17:30:00        Con heridos    4        09   
270761  10/03/2017 16:30:00        Con heridos    3        11   
270762  10/03/2017 17:10:00         Solo daños    3        05   
270763  10/03/2017 20:00:00        Con heridos    3        06   
270764  10/03/2017 22:40:00         Solo daños    3        05   

                           BARRIO                         LOCATION  
0                    Barrio Colón  [-75.5688011014, 6.24312304123]  
1                       San Diego  [-75.5693883283, 6.22870030622]  
2                      La Francia  [-75.5503911403, 6.29869502502]  
3       Santo Domingo Savio No. 1  [-75.5442195069, 6.30019383306]  
4                        Santa Fé   [-75.578180998, 6.21699243701]  
...                           ...                              ...  
270760                     Loreto  [-75.5559460731, 6.23260034372]  
270761             Cuarta Brigada  [-75.5851522144, 6.26044928627]  
270762      Francisco Antonio Zea  [-75.5713587099, 6.28457841509]  
270763                    Kennedy  [-75.5831132045, 6.28957484792]  
270764          Héctor Abad Gómez  [-75.5581989157, 6.30346727931]  

[270765 rows x 10 columns]

In [109]:
def extraer_dia(fecha):
    try:
        dia = int(fecha.split('/')[0][:2])
        return dia
    except:
        return None
def extraer_longitud(location):
    try:
        longitud = float(location.split(',')[0].replace('[', ''))
        return longitud
    except:
        return None
def extraer_latitud(location):
    try:
        latitud = float(location.split(',')[1].replace(']', ''))
        return latitud
    except:
        return None
#Extraemos el día, la longitud y la latitud
data['DIA'] = data['FECHA_ACCIDENTE'].apply(extraer_dia)
data['longitud'] = data['LOCATION'].apply(extraer_longitud)
data['latitud'] = data['LOCATION'].apply(extraer_latitud)
data=data.drop(['LOCATION'],axis=1)

Para mayor flexibilidad se va a redefinir el nombre de las variables.

In [110]:
data=data.rename(columns={'AÑO':'Anio','CLASE_ACCIDENTE':'Tipo','DISEÑO':'Diseno','GRAVEDAD_ACCIDENTE':'Gravedad','MES':'Mes','NUMCOMUNA':'NroComuna','BARRIO':'Barrio','DIA':'Dia','FECHA_ACCIDENTE':'Fecha_exacta'})
data.head()

Anio            Tipo      DIRECCION        Diseno         Fecha_exacta  \
0  2015  Caida Ocupante    CR 46 CL 43  Tramo de via  21/10/2015 05:58:00   
1  2015          Choque  CR 43 A CL 29  Tramo de via  05/11/2015 03:00:00   
2  2015            Otro    CR 46 CL 37  Tramo de via  21/10/2015 07:40:00   
3  2015       Atropello   CR 34 CL 112  Tramo de via  21/10/2015 12:00:00   
4  2015          Choque    CR 50 CL 14  Tramo de via  03/11/2015 06:16:00   

      Gravedad  Mes NroComuna                     Barrio  Dia   longitud  \
0  Con heridos   10        10               Barrio Colón   21 -75.568801   
1   Solo daños   11        10                  San Diego    5 -75.569388   
2  Con heridos   10        02                 La Francia   21 -75.550391   
3  Con heridos   10        01  Santo Domingo Savio No. 1   21 -75.544220   
4   Solo daños   11        15                   Santa Fé    3 -75.578181   

    latitud  
0  6.243123  
1  6.228700  
2  6.298695  
3  6.300194  
4  6.216992

Pasamos a ver que tipo de dato, está almacenando cada variable, para verificar que tena sentido en la manipulación.

In [111]:
data.dtypes

Anio             object
Tipo             object
DIRECCION        object
Diseno           object
Fecha_exacta     object
Gravedad         object
Mes               int64
NroComuna        object
Barrio           object
Dia               int64
longitud        float64
latitud         float64
dtype: object

Como podemos apreciar tenemos variables como el año, que sería más factibles si se trabajan como datos de tipo entero, porque así es más sencillo la manipulación de las fechas. Por lo que se va ver cuales son los valores diferentes que hay en la variable.

In [112]:
data['Anio'].unique()

array(['2015', '2014', '2016', '2017', '2018', '2019\\r', '2019', '2020'],
      dtype=object)

Como podemos apreciar hay un valor para los años '2019\\r' que debe hacer referencia al año 2019, para poder pasar los valores de los datos a enteros, vamos primero hacer la transformación a '2019', y luego de toda la columna a valores enteros.

In [113]:
# Reemplazamos '2019\\r' por '2019'
data['Anio'] = data['Anio'].str.strip()
data['Anio'] = data['Anio'].replace('2019\\r', '2019')
#Hacemos las conversión a valores enteros
data['Anio'] = data['Anio'].str.replace("'", '').astype(int)

Identificación de datos nulos

In [114]:
data.isnull().sum()

Anio                0
Tipo                6
DIRECCION           0
Diseno           1148
Fecha_exacta        0
Gravedad            0
Mes                 0
NroComuna           0
Barrio          19006
Dia                 0
longitud            0
latitud             0
dtype: int64

Los valores faltantes de la variable Diseno y la variable Tipo se van a inputar con la moda.

In [115]:
data['Tipo'] = data['Tipo'].fillna(data['Tipo'].mode()[0])
data['Diseno'] = data['Diseno'].fillna(data['Diseno'].mode()[0])

Como se puede apreciar a simple vista hay 19006 datos faltantes del barrio, pero vemos que en el resto de campos están todos los valores, por lo que se podrían llegar a inputar a partir de la dirección, vamos a ver más a fondo que contienen estos datos.

In [116]:
data[(data['Barrio'].isnull())]

Anio       Tipo                DIRECCION         Diseno  \
24      2015     Choque             TV 39 B CL 4   Tramo de via   
26      2015     Choque              CR 80 CL 15   Tramo de via   
36      2015     Choque             CL 647 CR 62   Tramo de via   
53      2015       Otro            CR 93 CL 97 D   Tramo de via   
69      2015     Choque  CL 33 A NorOrienteCR 64   Tramo de via   
...      ...        ...                      ...            ...   
270703  2017     Choque              CL 58 CR 96   Tramo de via   
270713  2017     Choque              CL 446 CR 1   Tramo de via   
270722  2017  Atropello            CL 999 CR 999  Lote o Predio   
270745  2017     Choque            CR 999 CL 999   Tramo de via   
270751  2017       Otro              CR 63 CL 91   Tramo de via   

               Fecha_exacta     Gravedad  Mes NroComuna Barrio  Dia  \
24      07/11/2015 09:00:00   Solo daños   11   Sin Inf    NaN    7   
26      09/11/2015 01:30:00  Con heridos   11   Sin Inf    NaN    9   
36      09/11/2015 06:20:00   Solo daños   11   Sin Inf    NaN    9   
53      10/11/2015 12:00:00  Con heridos   11   Sin Inf    NaN   10   
69      10/11/2015 03:25:00   Solo daños   11   Sin Inf    NaN   10   
...                     ...          ...  ...       ...    ...  ...   
270703  03/03/2017 07:20:00  Con heridos    3   Sin Inf    NaN    3   
270713  04/03/2017 12:28:00   Solo daños    3   Sin Inf    NaN    4   
270722  05/03/2017 23:00:00  Con heridos    3   Sin Inf    NaN    5   
270745  08/03/2017 09:54:00   Solo daños    3   Sin Inf    NaN    8   
270751  09/03/2017 14:30:00  Con heridos    3   Sin Inf    NaN    9   

         longitud   latitud  
24     -75.703776  6.221415  
26     -75.703776  6.221415  
36     -75.703776  6.221415  
53     -75.703776  6.221415  
69     -75.703776  6.221415  
...           ...       ...  
270703 -75.703776  6.221415  
270713 -75.703776  6.221415  
270722 -75.703776  6.221415  
270745 -75.703776  6.221415  
270751 -75.703776  6.221415  

[19006 rows x 12 columns]

Como se puede apreciar en los valores donde no se tiene el barrio, tampoco se tiene información de la comuna, y vemos que en los valores de longitud y latitud, la gran mayoría tienen el valor de -75.703776 y 6.221415, que corresponden a una ubicación totalmente alejada de la ciudad de Medellín por lo que los registros que tengan estos valores en estos dos campos se van a descartar ya que no nos sirven al para el propósito de ubicación en la aplicación, luego con los registros restantes en donde los valores sean diferentes se van intentar inputar mediante la API de gogle maps, a partir de la dirección, y los que no encuentre la aplicación simplmente se van a descartar.

In [117]:
data = data.drop(data[(data['longitud'] == -75.7037762763) & (data['latitud'] ==  6.22141524356)].index)
data[data['Barrio'].isnull()]

Anio    Tipo          DIRECCION        Diseno         Fecha_exacta  \
154     2015  Choque       Tramo 8 KM 9  Tramo de via  22/11/2015 09:15:00   
274     2015  Choque   Santa Elena KM 3  Tramo de via  07/12/2015 02:40:00   
288     2015  Choque       Tramo 8 KM 1  Tramo de via  09/12/2015 04:00:00   
390     2015  Choque    CL 12 sur CR 50  Tramo de via  03/11/2015 07:00:00   
1227    2015  Choque    CL 12 Sur CR 50  Tramo de via  20/10/2015 09:00:00   
...      ...     ...                ...           ...                  ...   
268845  2016  Choque    CL 12 Sur CR 49  Interseccion  09/12/2016 12:30:00   
268950  2016  Choque    CL 12 Sur CR 49  Tramo de via  23/12/2016 14:10:00   
269109  2015  Choque  Santa Elena KM 14  Tramo de via  25/11/2015 05:15:00   
269309  2015    Otro    CL 12 sur CR 50  Tramo de via  04/11/2015 03:00:00   
269490  2015  Choque    CL 12 sur CR 50  Tramo de via  23/11/2015 07:00:00   

           Gravedad  Mes NroComuna Barrio  Dia   longitud   latitud  
154      Solo daños   11   Sin Inf    NaN   22 -75.540183  6.153014  
274      Solo daños   12   Sin Inf    NaN    7 -75.529967  6.227976  
288     Con heridos   12   Sin Inf    NaN    9 -75.545148  6.206222  
390      Solo daños   11   Sin Inf    NaN    3 -75.581875  6.195397  
1227     Solo daños   10   Sin Inf    NaN   20 -75.581875  6.195397  
...             ...  ...       ...    ...  ...        ...       ...  
268845   Solo daños   12   Sin Inf    NaN    9 -75.581106  6.195185  
268950   Solo daños   12   Sin Inf    NaN   23 -75.581106  6.195185  
269109  Con heridos   11   Sin Inf    NaN   25 -75.501664  6.219353  
269309  Con heridos   11   Sin Inf    NaN    4 -75.581875  6.195397  
269490   Solo daños   11   Sin Inf    NaN   23 -75.581875  6.195397  

[440 rows x 12 columns]

In [118]:
from geopy.geocoders import GoogleV3
def obtener_barrio_medellin(direccion):
    query = direccion +', Medellín, Antioquia'
    geolocalizador = GoogleV3(api_key='AIzaSyCO_JBKNW8FXmtu0f1APiMUHd1KtxuiXJo')
    try:
        ubicacion = geolocalizador.geocode(query)
        if ubicacion:
            componentes = ubicacion.raw['address_components']
            for componente in componentes:
              if 'neighborhood' in componente['types'] or 'sublocality' in componente['types']:
                  return componente['long_name']
    except Exception as e:
        print("Error al obtener la ubicación:", e)
    return None

In [119]:
data['Barrio'].fillna(data.loc[data['Barrio'].isnull()]['DIRECCION'].apply(obtener_barrio_medellin), inplace=True)

In [120]:
data = data.dropna(subset=['Barrio'])
data.isnull().sum()

Anio            0
Tipo            0
DIRECCION       0
Diseno          0
Fecha_exacta    0
Gravedad        0
Mes             0
NroComuna       0
Barrio          0
Dia             0
longitud        0
latitud         0
dtype: int64

De igual forma se van a eliminar los registros duplicados, ya que no tiene mucho sentido de que se tenga registro del mismo accidente más de una vez, con la misma fecha, a la misma hora, en la mimsa ubicación geográfica.

In [121]:
data.duplicated().sum()

180

In [122]:
data = data.drop_duplicates()
data

Anio            Tipo      DIRECCION         Diseno  \
0       2015  Caida Ocupante    CR 46 CL 43   Tramo de via   
1       2015          Choque  CR 43 A CL 29   Tramo de via   
2       2015            Otro    CR 46 CL 37   Tramo de via   
3       2015       Atropello   CR 34 CL 112   Tramo de via   
4       2015          Choque    CR 50 CL 14   Tramo de via   
...      ...             ...            ...            ...   
270760  2017       Atropello  CR 28 CL 38 F  Lote o Predio   
270761  2017  Caida Ocupante    CL 53 CR 69  Lote o Predio   
270762  2017          Choque  CR 64 C CL 92   Tramo de via   
270763  2017  Caida Ocupante  CR 78 A CL 92   Tramo de via   
270764  2017          Choque   CR 63 CL 110   Tramo de via   

               Fecha_exacta     Gravedad  Mes NroComuna  \
0       21/10/2015 05:58:00  Con heridos   10        10   
1       05/11/2015 03:00:00   Solo daños   11        10   
2       21/10/2015 07:40:00  Con heridos   10        02   
3       21/10/2015 12:00:00  Con heridos   10        01   
4       03/11/2015 06:16:00   Solo daños   11        15   
...                     ...          ...  ...       ...   
270760  21/04/2017 17:30:00  Con heridos    4        09   
270761  10/03/2017 16:30:00  Con heridos    3        11   
270762  10/03/2017 17:10:00   Solo daños    3        05   
270763  10/03/2017 20:00:00  Con heridos    3        06   
270764  10/03/2017 22:40:00   Solo daños    3        05   

                           Barrio  Dia   longitud   latitud  
0                    Barrio Colón   21 -75.568801  6.243123  
1                       San Diego    5 -75.569388  6.228700  
2                      La Francia   21 -75.550391  6.298695  
3       Santo Domingo Savio No. 1   21 -75.544220  6.300194  
4                        Santa Fé    3 -75.578181  6.216992  
...                           ...  ...        ...       ...  
270760                     Loreto   21 -75.555946  6.232600  
270761             Cuarta Brigada   10 -75.585152  6.260449  
270762      Francisco Antonio Zea   10 -75.571359  6.284578  
270763                    Kennedy   10 -75.583113  6.289575  
270764          Héctor Abad Gómez   10 -75.558199  6.303467  

[250863 rows x 12 columns]

Se va eliminar la variable Fecha_exacta ya que lo elementos de interés que son el día el año y el mes ya se tienen, se uso básicamente para corroborar valores duplicados, ya que la hora exacta no permitia verificar con mayor seguridad si dos registros eran iguales.

In [123]:
data= data.drop(['Fecha_exacta'],axis=1)

Con esta última limpieza vamos a cerciorarnos de que todos los valores en las variables tengan sentido, y además en las variables categóricas vamos a ver como se distribuyen las distintas categórias.



***Barrio***

In [124]:
len(data['Barrio'].unique())

492

En los valores del Barrio hay varios errores, en primer lugar según la Alcaldía de Medellín se cuenta con 249 barrios urbanos oficiales, 52 veredas, 20 areas áreas Institucionales, 7 áreas de expansión (urbano – rural). Pero vemos que la cantidad de datos únicos de los barrios son 492, por lo que analizando a profundidad los valores vemos que muchos están por código de barrio por lo que se van modificar con su respectivo barrio, y también vemos los caracteres que anteriormente se mencionaron, como lo son las letras con tildes y las ñ, por lo que también se van a modificar.

In [ ]:
data['Barrio'].unique()

In [126]:
barrios = pd.read_csv('/content/drive/Shareddrives/FdeAnaliticaG4/Proyecto2/barrios.csv', sep=',', low_memory=False)
barrios = barrios.set_index('codigo')['barrio'].to_dict()
barrios = {str(k).zfill(4): v for k, v in barrios.items()}
for key, value in barrios.items():
    data['Barrio'] = data['Barrio'].str.replace(key, value)

In [127]:
def corregir_caracteres(texto):
    caracteres = {
        "\\xE1": "á",
        "\\xE9": "é",
        "\\xED": "í",
        "\\xF3": "ó",
        "\\xFA": "ú",
        "\\xF1": "ñ",
        "\\xC1": "Á",
        "\\xC9": "É",
        "\\xCD": "Í",
        "\\xD3": "Ó",
        "\\xDA": "Ú",
        "\\xD1": "Ñ",
        "No. 2":"No.2",
        "No. 1":"No.1",
        "No. 3":"No.3",
        'Villa Lilliam':'Villa Liliam',
        'Simón Bolivar':'Simón Bolívar',
        'Santa María de Los Ángeles':'Santa María de los Ángeles',
        'Nueva Villa de la Iguaná':'Nueva Villa de La Iguaná',
        'El Nogal - Los Almendros':'El Nogal Los Almendros',
        'Berlin':'Berlín',
        'Barrios de Jesús': 'Barrio de Jesús',
        'B. Cerro  El Volador':'B. Cerro El Volador',
        'Aguas frias':'Aguas Frias',
        'Sin Nombre':'El Pesebre',
        'Inst':'Laureles',
        'AUC1':'San Cristóbal cabecera',
        '0':'La Aguacatala',
        'El Nogal Los Almendros':'El Nogal-Los Almendros',
        '9La Aguacatala86':'La Aguacatala',
                  'Barrio Caycedo':'Barrio Caicedo',
                            'SUBURB El Plan':'Suburbano El Plan',
                            'AUC2':'El Astillero'
    }
    for key, value in caracteres.items():
        texto = texto.replace(key, value)
    return texto

# Aplica la función a la columna "Barrio" para corregir los caracteres
data['Barrio'] = data['Barrio'].apply(corregir_caracteres)

In [ ]:
sorted(list(data['Barrio'].unique()))

In [129]:
len(data['Barrio'].unique())

328

Con esto obtenemos un total de 328 valores diferentes para los barrios, donde están incluidos veredas, áreas de expansión o en algunos casos territorios que pueden estar alejados de la ciudad.

***NroComuna***

In [ ]:
data['NroComuna'].value_counts().sort_index()

En este caso observamos que como el tipo de dato de la variable es un object, tiene valores que representan los mismo como lo es '01' y '1' y así para los valores hasta el 9. por lo que se van agrupar en uno solo. Vemos además que hay valores como 'AU', 'In', 'SN', 'Sin Inf', '0' que no están representando un valor adecuado para la comuna, recordemos que Medellín cuenta con 16 comunas urbanas y 5 comunas rurales (50, 60, 70, 80 , 90). Por lo que se van a determinar el valor  verdadero de la comuna a partir del barrio.

In [131]:
def corregir_cmna(texto):
    caracteres = {
        '01': '1',
        '02': '2',
        '03': '3',
        '04': '4',
        '05': '5',
        '06': '6',
        '07': '7',
        '08': '8',
        '09': '9'
    }
    for key, value in caracteres.items():
        texto = texto.replace(key, value)
    return texto

# Aplica la función a la columna "Barrio" para corregir los caracteres
data['NroComuna'] = data['NroComuna'].apply(corregir_cmna)

In [132]:
df_correccion = data.loc[data['NroComuna'].isin(['AU', 'Sin Inf','SN','In','0'])]
df_correccion

Anio    Tipo        DIRECCION         Diseno     Gravedad  Mes  \
154     2015  Choque     Tramo 8 KM 9   Tramo de via   Solo daños   11   
288     2015  Choque     Tramo 8 KM 1   Tramo de via  Con heridos   12   
390     2015  Choque  CL 12 sur CR 50   Tramo de via   Solo daños   11   
773     2014  Choque    CL 47 D CR 70   Tramo de via   Solo daños   10   
1157    2014    Otro  CL 53 D CR 85 E   Tramo de via  Con heridos   10   
...      ...     ...              ...            ...          ...  ...   
268819  2016  Choque    CL 47 D CR 70  Lote o Predio   Solo daños   12   
268845  2016  Choque  CL 12 Sur CR 49   Interseccion   Solo daños   12   
268950  2016  Choque  CL 12 Sur CR 49   Tramo de via   Solo daños   12   
269309  2015    Otro  CL 12 sur CR 50   Tramo de via  Con heridos   11   
269490  2015  Choque  CL 12 sur CR 50   Tramo de via   Solo daños   11   

       NroComuna      Barrio  Dia   longitud   latitud  
154      Sin Inf        Mesa   22 -75.540183  6.153014  
288      Sin Inf        Mesa    9 -75.545148  6.206222  
390      Sin Inf    Guayabal    3 -75.581875  6.195397  
773           In    Laureles   17 -75.587903  6.253172  
1157          SN  El Pesebre   28 -75.601042  6.270381  
...          ...         ...  ...        ...       ...  
268819        In    Laureles    5 -75.587903  6.253172  
268845   Sin Inf  El Poblado    9 -75.581106  6.195185  
268950   Sin Inf  El Poblado   23 -75.581106  6.195185  
269309   Sin Inf    Guayabal    4 -75.581875  6.195397  
269490   Sin Inf    Guayabal   23 -75.581875  6.195397  

[446 rows x 11 columns]

In [134]:
comunas_barrios = {}
for clave, barrio in barrios.items():
    comuna = clave[:2]  # Obtiene los primeros dos dígitos de la clave
    if comuna[0] == '0':
        comuna = comuna[1]
    if comuna not in comunas_barrios:
        comunas_barrios[comuna] = []  # Crea una lista vacía si la comuna no existe
    comunas_barrios[comuna].append(barrio)
df_correccion['NroComuna'] = df_correccion['Barrio'].map({barrio: comuna for comuna, barrios in comunas_barrios.items() for barrio in barrios})
df_correccion

Anio    Tipo        DIRECCION         Diseno     Gravedad  Mes  \
154     2015  Choque     Tramo 8 KM 9   Tramo de via   Solo daños   11   
288     2015  Choque     Tramo 8 KM 1   Tramo de via  Con heridos   12   
390     2015  Choque  CL 12 sur CR 50   Tramo de via   Solo daños   11   
773     2014  Choque    CL 47 D CR 70   Tramo de via   Solo daños   10   
1157    2014    Otro  CL 53 D CR 85 E   Tramo de via  Con heridos   10   
...      ...     ...              ...            ...          ...  ...   
268819  2016  Choque    CL 47 D CR 70  Lote o Predio   Solo daños   12   
268845  2016  Choque  CL 12 Sur CR 49   Interseccion   Solo daños   12   
268950  2016  Choque  CL 12 Sur CR 49   Tramo de via   Solo daños   12   
269309  2015    Otro  CL 12 sur CR 50   Tramo de via  Con heridos   11   
269490  2015  Choque  CL 12 sur CR 50   Tramo de via   Solo daños   11   

       NroComuna      Barrio  Dia   longitud   latitud  
154          NaN        Mesa   22 -75.540183  6.153014  
288          NaN        Mesa    9 -75.545148  6.206222  
390           15    Guayabal    3 -75.581875  6.195397  
773           11    Laureles   17 -75.587903  6.253172  
1157          13  El Pesebre   28 -75.601042  6.270381  
...          ...         ...  ...        ...       ...  
268819        11    Laureles    5 -75.587903  6.253172  
268845        14  El Poblado    9 -75.581106  6.195185  
268950        14  El Poblado   23 -75.581106  6.195185  
269309        15    Guayabal    4 -75.581875  6.195397  
269490        15    Guayabal   23 -75.581875  6.195397  

[446 rows x 11 columns]

In [136]:
data.update(df_correccion)

In [139]:
data[['Anio','Mes','Dia']] = data[['Anio','Mes','Dia']].astype(int)

Anio           int64
Tipo          object
DIRECCION     object
Diseno        object
Gravedad      object
Mes            int64
NroComuna     object
Barrio        object
Dia            int64
longitud     float64
latitud      float64
dtype: object

In [140]:
data.loc[data['NroComuna'].isin(['AU', 'Sin Inf','SN','In','0'])]['Barrio'].unique()

array(['Mesa', 'Zona 9'], dtype=object)

Vemos que quedaron faltando las comunas para los barrios Mesa y Zona que son barrios que pertenecen a Envigado, por lo que se van a eliminar estos registros.

In [141]:
data = data[~data['Barrio'].isin(['Mesa','Zona 9'])]

In [ ]:
data['NroComuna'].value_counts().sort_index()

Eliminamos la columna DIRECCION que para el propósito no es tan relevante.

In [143]:
data = data.drop(columns='DIRECCION',axis=1)

***Año***

In [145]:
data['Anio'].value_counts()

2016    43517
2017    42706
2015    42689
2019    42422
2018    40202
2014    20756
2020    18440
Name: Anio, dtype: int64

***Tipo de accidente***

In [146]:
data['Tipo'].value_counts()

Choque               169381
Otro                  26691
Atropello             24036
Caida Ocupante        15747
Volcamiento            9051
Caída de Ocupante      5783
Incendio                 28
Caída Ocupante           15
Name: Tipo, dtype: int64

Como se puede apreciar, en la variable Tipo se tienen 3 categorías que siginifican los mismo pero están escritas de manera distina: Caida Ocupante, Caída de Ocupante, Caída Ocupante. Por lo que se van a reunir en una sola.

In [147]:
data['Tipo'] = data['Tipo'].replace(['Caída de Ocupante', 'Caída Ocupante'], 'Caida Ocupante')
data['Tipo'].value_counts()

Choque            169381
Otro               26691
Atropello          24036
Caida Ocupante     21545
Volcamiento         9051
Incendio              28
Name: Tipo, dtype: int64

***Diseño***

In [148]:
data['Diseno'].value_counts()

Tramo de via     173895
Interseccion      42329
Lote o Predio     24019
Glorieta           6354
Paso Elevado       1315
Puente              967
Ciclo Ruta          963
Paso Inferior       643
Paso a Nivel        113
Via peatonal         54
Tunel                51
Pontón               23
Pont\xF3n             6
Name: Diseno, dtype: int64

En este caso vemos que hay un error en la categoria Pont\xF3n,  donde "\xF3" representa un carácter especial en la codificación Latin-1 (ISO 8859-1) que corresponde a la letra "ó" con acento. Por lo que se va cambiar por Ponton, de igual modo le quitamos la tilde a Pontón.

In [149]:
data['Diseno'] = data['Diseno'].replace(['Pont\\xF3n','Pontón'], 'Ponton')
data['Diseno'].value_counts()

Tramo de via     173895
Interseccion      42329
Lote o Predio     24019
Glorieta           6354
Paso Elevado       1315
Puente              967
Ciclo Ruta          963
Paso Inferior       643
Paso a Nivel        113
Via peatonal         54
Tunel                51
Ponton               29
Name: Diseno, dtype: int64

***Gravedad***

In [150]:
data['Gravedad'].value_counts()

Con heridos      137228
Solo daños        92883
Solo da\xF1os     19238
Con muertos        1383
Name: Gravedad, dtype: int64

En este caso también vemos una error en la codificación de caracteres nuevamente, donde \xF1 correponde a la letra ñ en varias codificaciones  incluyendo Latin-1 (ISO 8859-1), por lo que se van a cambiar ambos valores por 'Solo danos'.

In [151]:
data['Gravedad']= data['Gravedad'].replace(['Solo daños','Solo da\\xF1os'], 'Solo danos')
data['Gravedad'].value_counts()

Con heridos    137228
Solo danos     112121
Con muertos      1383
Name: Gravedad, dtype: int64

***Mes***

In [152]:
data['Mes'].value_counts()

8     25095
7     23263
9     22416
10    22252
12    21503
11    20713
2     20661
3     19876
5     19669
6     18813
1     18609
4     17862
Name: Mes, dtype: int64

***Dia***

In [153]:
data['Dia'].value_counts().sort_index()

1     7907
2     8062
3     8072
4     8262
5     8491
6     8452
7     8066
8     8053
9     8165
10    8510
11    8335
12    8363
13    8283
14    8210
15    8304
16    8739
17    8464
18    8682
19    8248
20    8156
21    8482
22    8324
23    8436
24    8034
25    7935
26    8155
27    8275
28    8240
29    7465
30    7185
31    4377
Name: Dia, dtype: int64

In [387]:
data.to_csv('/content/drive/Shareddrives/FdeAnaliticaG4/Proyecto2/DatosHistoricos.csv', index=False)

Con esto se finaliza la limpieza de los datos, quedandonos con 250737 registros y asegurandonos que cada una de las variables contiene información valiosa para el desarrollo de los modelos y los agrupamientos.

# ***PRUEBAS***

In [ ]:
from geopy.geocoders import Nominatim

def obtener_informacion_geografica_por_coordenadas(latitud, longitud):
    geolocalizador = Nominatim(user_agent="myGeocoder")  # Proporciona un nombre de usuario único como "user_agent"

    try:
        ubicacion = geolocalizador.reverse((latitud, longitud))
        if ubicacion:
            return ubicacion.raw
    except Exception as e:
        print("Error al obtener la ubicación:", e)

    return None

# Ejemplo de uso
latitud = 6.30019383306  # Reemplaza con la latitud deseada
longitud = -75.5442195069 # Reemplaza con la longitud deseada

informacion = obtener_informacion_geografica_por_coordenadas(latitud, longitud)

if informacion:
    print(informacion)

{'place_id': 300172711, 'licence': 'Data © OpenStreetMap contributors, ODbL 1.0. http://osm.org/copyright', 'osm_type': 'way', 'osm_id': 608944935, 'lat': '6.3001395', 'lon': '-75.5439745', 'class': 'highway', 'type': 'residential', 'place_rank': 26, 'importance': 0.10000999999999993, 'addresstype': 'road', 'name': 'Carrera 33', 'display_name': 'Carrera 33, Santo Domingo Savio No.1, Comuna 1 - Popular, Perímetro Urbano Medellín, Medellín, Valle de Aburrá, Antioquia, RAP del Agua y la Montaña, 005004, Colombia', 'address': {'road': 'Carrera 33', 'neighbourhood': 'Santo Domingo Savio No.1', 'suburb': 'Comuna 1 - Popular', 'city': 'Perímetro Urbano Medellín', 'state_district': 'Valle de Aburrá', 'state': 'Antioquia', 'ISO3166-2-lvl4': 'CO-ANT', 'region': 'RAP del Agua y la Montaña', 'postcode': '005004', 'country': 'Colombia', 'country_code': 'co'}, 'boundingbox': ['6.2997543', '6.3001850', '-75.5439770', '-75.5439152']}


In [ ]:
# Crear un nuevo DataFrame con los valores únicos por columna
valores_unicos_por_columna = pd.DataFrame({
    'Variable': data.columns,
    'Valores Únicos': [data[col].unique() for col in data.columns]
})
valores_unicos_por_columna

Variable                                     Valores Únicos
0       Anio         [2015, 2014, 2016, 2017, 2018, 2019, 2020]
1       Tipo  [Caida Ocupante, Choque, Otro, Atropello, Volc...
2     Diseno  [Tramo de via, Interseccion, Glorieta, Lote o ...
3   Gravedad             [Con heridos, Solo danos, Con muertos]
4        Mes            [10, 11, 12, 9, 4, 3, 5, 6, 7, 8, 1, 2]
5  NroComuna  [10, 2, 1, 15, 14, 11, 7, 6, 4, 9, 5, 80, 13, ...
6     Barrio  [Barrio Colón, San Diego, La Francia, Santo Do...
7        Dia  [21, 5, 3, 18, 4, 22, 6, 7, 8, 9, 10, 11, 12, ...
8   longitud  [-75.5688011014, -75.5693883283, -75.550391140...
9    latitud  [6.24312304123, 6.22870030622, 6.29869502502, ...

In [79]:
from geopy.geocoders import GoogleV3

def obtener_barrio_medellin(direccion, cache={}):
    # Revisa si el resultado ya está en el caché
    if direccion in cache:
        return cache[direccion]

    query = direccion + ', Medellín, Antioquia'
    geolocalizador = GoogleV3(api_key='AIzaSyCO_JBKNW8FXmtu0f1APiMUHd1KtxuiXJo')  # Reemplaza con tu propia clave de API de Google

    try:
        ubicacion = geolocalizador.geocode(query)
        if ubicacion:
            componentes = ubicacion.raw['address_components']
            for componente in componentes:
                if 'neighborhood' in componente['types'] or 'sublocality' in componente['types']:
                    barrio = componente['long_name']
                    # Guarda el resultado en el caché
                    cache[direccion] = barrio
                    return barrio
    except Exception as e:
        print("Error al obtener la ubicación:", e)

    # En caso de error o falta de datos, guarda None en el caché
    cache[direccion] = None
    return None

print(obtener_barrio_medellin('CR 72 CL 404'))

None


In [106]:
dataprueba = datanull.head(10)
dataprueba['Barrio'] = dataprueba['DIRECCION'].apply(obtener_barrio_medellin)
dataprueba

<ipython-input-106-3159ffbf75c1>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataprueba['Barrio'] = dataprueba['DIRECCION'].apply(obtener_barrio_medellin)


Anio            Tipo                DIRECCION         Diseno  \
24   2015          Choque             TV 39 B CL 4   Tramo de via   
26   2015          Choque              CR 80 CL 15   Tramo de via   
36   2015          Choque             CL 647 CR 62   Tramo de via   
53   2015            Otro            CR 93 CL 97 D   Tramo de via   
69   2015          Choque  CL 33 A NorOrienteCR 64   Tramo de via   
95   2015  Caida Ocupante             CR 72 CL 404   Tramo de via   
138  2015          Choque          CL 16 B CR 25 B  Lote o Predio   
154  2015          Choque             Tramo 8 KM 9   Tramo de via   
162  2015          Choque            CR 999 CL 999   Tramo de via   
176  2015       Atropello            CL 62 CR 18 A   Tramo de via   

            Fecha_exacta     Gravedad  Mes NroComuna              Barrio  \
24   07/11/2015 09:00:00   Solo daños   11   Sin Inf  Laureles - Estadio   
26   09/11/2015 01:30:00  Con heridos   11   Sin Inf                None   
36   09/11/2015 06:20:00   Solo daños   11   Sin Inf       La Candelaria   
53   10/11/2015 12:00:00  Con heridos   11   Sin Inf                None   
69   10/11/2015 03:25:00   Solo daños   11   Sin Inf  Laureles - Estadio   
95   17/11/2015 07:00:00  Con heridos   11   Sin Inf                None   
138  21/11/2015 04:00:00   Solo daños   11   Sin Inf                None   
154  22/11/2015 09:15:00   Solo daños   11   Sin Inf                Mesa   
162  23/11/2015 11:50:00  Con heridos   11   Sin Inf         El Castillo   
176  26/11/2015 04:30:00  Con heridos   11   Sin Inf                None   

            X           Y  Dia   longitud   latitud  
24   820000.0  1180000.00    7 -75.703776  6.221415  
26   820000.0  1180000.00    9 -75.703776  6.221415  
36   820000.0  1180000.00    9 -75.703776  6.221415  
53   820000.0  1180000.00   10 -75.703776  6.221415  
69   820000.0  1180000.00   10 -75.703776  6.221415  
95   820000.0  1180000.00   17 -75.703776  6.221415  
138  820000.0  1180000.00   21 -75.703776  6.221415  
154  838090.2  1172380.35   22 -75.540183  6.153014  
162  820000.0  1180000.00   23 -75.703776  6.221415  
176  820000.0  1180000.00   26 -75.703776  6.221415